app/javascript/lib/validation/validate-remote-file-content.js

```diff
@@ -0,0 +1,77 @@
+/**
+* @fileoverview Client-side file validation (CSFV) for sync
+*/
+
+import { validateFileContent } from './validate-file-content'
+import { fetchBucketFile } from 'lib/scp-api'
+
+/**
+ * 50 MiB, max number of bytes to read and validate from remote file for sync.
+ * Median fixed US bandwidth is 16 MiB/s (17 MB/s, 136 Mbps) as of 2021-12
+ * per https://www.speedtest.net/global-index/united-states.  > 80% of ingests
+ * are for files <= 50 MiB per https://mixpanel.com/s/3EbFMj.
+ *
+ * So 50 MiB means sync CSFV fully scans > 80% files, usually in < 5 seconds.
+ * Local tests gave 4.6 s (3.4 s remote read, 1.1 s validate; 138 Mbps down).
+ */
+export const MAX_SYNC_CSFV_BYTES = 50 * 1024 * 1024
+
+/** Add file-size-related entries to syncProps */
+export function updateSyncProps(contentRange, contentLength, syncProps) {
+  // Total size of the file in bytes
+  const fileSizeTotal = parseInt(contentRange.split('/')[1])
+
+  // Total bytes downloaded, which can be much less than total file size
+  const fileSizeFetched = parseInt(contentLength)
+
+  const fetchedCompleteFile = (fileSizeTotal === fileSizeFetched)
+  Object.assign(syncProps, {
+    fileSizeFetched,
+    fileSizeTotal,
+    fetchedCompleteFile
+  })
+
+  return syncProps
+}
+
+/** Construct a File object from superficially-parsed range response data */
+export function getFileFromRangeData(content, fileName, contentType, syncProps) {
+  // If Range request didn't fetch the full file, then truncate the last
+  // line, which is almost certainly incomplete and thus invalid.
+  let cleanLineContent = content
+  if (!syncProps.fetchedCompleteFile) {
+    cleanLineContent = content.split('\n').slice(0, -1).join('\n')
+  }
+
+  const file = new File([cleanLineContent], fileName, { type: contentType })
+
+  return file
+}
+
+/**
+* Validate a file in a GCS bucket; used for sync
+*/
+export async function validateRemoteFileContent(
+  bucketName, fileName, fileType, fileOptions
+) {
```

Comment:
The perftime measurement and other features will be simplified if this function is a very thin wrapper around `validateFileContent`.  All it needs to do is call validateFileContent with the bucketName added to fileOptions.

Then put most of the code below into a new function `{file, fileInfo} = createFileFromBucket(fileName, bucketName, maxBytes)` that returns the file object and `fileInfo`, which is  whatever metainfo (e.g. perfTimes) you need.  That way the everything lives inside the main timing loop, and `createFileFromBucket` is a more independently testable/useful method.  Assuming `createFileFromBucket` gets called from inside `parseFile`, this also opens the door for easy optimizations like only fetching the first few bytes of non-parseable file types, since all we want to do is validate the gzip encoding

Reply:
Let's revisit this in a separate task, e.g. auto-selecting core file types for sync (SCP-4065).

A nice aspect of the original approach here is that validating remote files is structurally similar to validating local files.  In both local and remote cases, the `File` object is created outside `validateFileContent` (and thus outside `parseFile`).  Beyond the parallel structure, surely that's clearer than _creating_ a Foo in a function named `parseFoo`.

I like your point about validating gzip encoding.  I could see fetching the first ~500 kB instead of 3-4 bytes, then validating both magic numbers for gzip _and_ (assuming that succeeds) infer SCP file study type from the first few lines.  Perhaps both could be part of a pre-emptive tier of validation done before the user selects the file type.  In any case, I think such work is best scoped for another ticket.

Reply:
nice, very much appreciate the new refactoring!